runtime/vm/ValueTypeHelpers.hpp

```diff
@@ -561,6 +561,101 @@ class VM_ValueTypeHelpers {
+	static VMINLINE I_32
+	copyFlattenableArray(J9VMThread *currentThread, MM_ObjectAccessBarrierAPI objectAccessBarrier, MM_ObjectAllocationAPI objectAllocate, j9object_t srcObject, j9object_t destObject, I_32 srcIndex, I_32 destIndex, I_32 lengthInSlots)
+	{
+		I_32 srcEndIndex = srcIndex + lengthInSlots;
+		bool typeChecksRequired = false;
+		J9Class *srcClazz = J9OBJECT_CLAZZ(currentThread, srcObject);
+		J9Class *destClazz = J9OBJECT_CLAZZ(currentThread, destObject);
+		J9Class *destComponentClass = ((J9ArrayClass *)destClazz)->componentType;
+
+		/*
+		* Decide if type checks are required
+		* Type checks are not required only if both classes are the same, or the source class is a subclass of the dest class
+		*/
+		if (srcClazz != destClazz) {
+			UDATA srcDepth = J9CLASS_DEPTH(srcClazz);
+			UDATA destDepth = J9CLASS_DEPTH(destClazz);
+			if ((srcDepth <= destDepth) || (destClazz->superclasses[srcDepth] != srcClazz)) {
+				typeChecksRequired = true;
+			}
+		}
+
+		/* Array elements must be copied backwards if source and destination overlap in memory and source is before destination */
+		if ((srcObject == destObject) && (srcIndex < destIndex) && ((srcIndex + lengthInSlots) > destIndex)) {
+			srcEndIndex = srcIndex;
+			srcIndex += lengthInSlots;
+			destIndex += lengthInSlots;
+
+			while (srcIndex > srcEndIndex) {
+				srcIndex--;
+				destIndex--;
+
+				j9object_t copyObject = loadFlattenableArrayElement(currentThread, objectAccessBarrier, objectAllocate, srcObject, srcIndex, true);
+
+				/* Check for an allocation failure (only possible if srcClazz is flattened) */
+				if ((NULL == copyObject) && J9_IS_J9CLASS_FLATTENED(srcClazz)) {
+					VM_VMHelpers::pushObjectInSpecialFrame(currentThread, srcObject);
+					copyObject = loadFlattenableArrayElement(currentThread, objectAccessBarrier, objectAllocate, srcObject, srcIndex, false);
+					srcObject = VM_VMHelpers::popObjectInSpecialFrame(currentThread);
+				}
+
+				/* No type checks required since srcObject == destObject */
+
+				storeFlattenableArrayElement(currentThread, objectAccessBarrier, destObject, destIndex, copyObject);
+			}
+		} else {
+			while (srcIndex < srcEndIndex) {
+				j9object_t copyObject = loadFlattenableArrayElement(currentThread, objectAccessBarrier, objectAllocate, srcObject, srcIndex, true);
+
+				/* Check for an allocation failure (only possible if srcClazz is flattened) */
+				if ((NULL == copyObject) && J9_IS_J9CLASS_FLATTENED(srcClazz)) {
+					VM_VMHelpers::pushObjectInSpecialFrame(currentThread, srcObject);
+					copyObject = loadFlattenableArrayElement(currentThread, objectAccessBarrier, objectAllocate, srcObject, srcIndex, false);
+					srcObject = VM_VMHelpers::popObjectInSpecialFrame(currentThread);
+				}
+
+				if (typeChecksRequired) {
+					if (!VM_VMHelpers::objectArrayStoreAllowed(currentThread, destObject, copyObject)) {
+						return -1;
+					}
+					if (J9_IS_J9CLASS_PRIMITIVE_VALUETYPE(destComponentClass) && (NULL == copyObject)) {
```

Comment:
I suggest to do the NPE check before the `objectArrayStoreAllowed()` check.

Reply:
moved in [25f3eed](https://github.com/eclipse-openj9/openj9/commit/25f3eed6d26349b79515af1233841c0436016b4e)